sdk/servicebus/Azure.Messaging.ServiceBus/tests/Infrastructure/ServiceBusScope.cs

```diff
@@ -0,0 +1,627 @@
+                    // test run will also remove the orphan.
+                }
+
+                _disposed = true;
+            }
+        }
+
+        /// <summary>
+        ///   Provides access to dynamically created Service Bus Topic instance which exists only in the context
+        ///   of the scope; disposal removes the resource from Azure.
+        /// </summary>
+        ///
+        /// <seealso cref="IAsyncDisposable" />
+        ///
+        public sealed class TopicScope : IAsyncDisposable
+        {
+            /// <summary>Serves as a sentinel flag to denote when the instance has been disposed.</summary>
+            private bool _disposed = false;
+
+            /// <summary>
+            ///   The name of the Service Bus namespace associated with the queue.
+            /// </summary>
+            ///
+            public string NamespaceName { get; }
+
+            /// <summary>
+            ///  The name of the topic.
+            /// </summary>
+            ///
+            public string TopicName { get; }
+
+            /// <summary>
+            ///   The set of names for the subscriptions associated with the topic.
+            /// </summary>
+            ///
+            public IReadOnlyList<string> SubscriptionNames { get; }
+
+            /// <summary>
+            ///   A flag indicating if the topic should be removed when the scope is complete.
+            /// </summary>
+            ///
+            /// <value><c>true</c> if the queue was should be removed at scope completion; otherwise, <c>false</c>.</value>
+            ///
+            private bool ShouldRemoveAtScopeCompletion { get; }
+
+            /// <summary>
+            ///   Initializes a new instance of the <see cref="TopicScope"/> class.
+            /// </summary>
+            ///
+            /// <param name="serviceBusNamespaceName">The name of the Service Bus namespace to which the queue is associated.</param>
+            /// <param name="topicName">The name of the topic.</param>
+            /// <param name="subscriptionNames">The set of names for the subscriptions </param>
+            /// <param name="shouldRemoveAtScopeCompletion">A flag indicating whether the topic should be removed when the scope is complete.</param>
+            ///
+            public TopicScope(string serviceBusNamespaceName,
+                              string topicName,
+                              IReadOnlyList<string> subscriptionNames,
+                              bool shouldRemoveAtScopeCompletion)
+            {
+                NamespaceName = serviceBusNamespaceName;
+                TopicName = topicName;
+                SubscriptionNames = subscriptionNames;
+                ShouldRemoveAtScopeCompletion = shouldRemoveAtScopeCompletion;
+            }
+
+            /// <summary>
+            ///   Performs the tasks needed to remove the dynamically created Service Bus Queue.
```

Comment:
nit: should mention topic

Any chance this and the queueScope code can be combined - seems the only difference is the name of the property on the management client and whether we pass a subscription.

Reply:
gah... yeah.  Queue is a typo.    The split allows you to access members specific to the type of scope that you're using as well as keeping the cleanup logic focused.   The counter-question that I'd pose is "what do you feel the benefit of having a meta class that combines two different entities would be?"

Reply:
It would be nice to avoid duplicating code so if things change we only have to update one spot. 
Also Queues and Topics really are both ServiceBusEntities so having an abstract class ServiceBusEntity that they derive from makes sense to me.

Reply:
What would you change, though, that is shared?    The dispose method name can't change and the implementation is entity specific.   The `NamespaceName` and `ShouldRemoveAtScopeCompletion` are really the only shared members that you could change, and both are simple property getters.  I'm going to take a YANGI approach here; if we find that we have common members that we want to share amongst entities and would have to add them multiple times, I think that's the point of reconsidering whether a base entity scope makes sense.

Reply:
The DisposeAsync implementations look identical except for the name of the property that holds the entity name and the type of the entity on the ServiceBusManagementClient. If you had an EntityName property on an abstract class I would think the entire method could be moved up one level and shared by both TopicScope and QueueScope. You would still have Delete methods on TopicScope/QueueScope that the abstract class would delegate to for that one line - `                    await CreateRetryPolicy().ExecuteAsync(() => client.Queues.DeleteAsync(resourceGroup, NamespaceName, QueueName)).ConfigureAwait(false);
`

Reply:
Fair, and a legit observation.  I think the point where we disagree is whether taking on additional complexity by introducing a class hierarchy offers greater benefit by removing duplication than it does hurting locality of information and creating a greater cognitive burden on those reading the code.  I find myself in the camp that believes we often are premature to introduce complexity.   In this particular case, I'm of the opinion that its worth trading that small bit of duplication of sparse, and simple boilerplate code that we're unlikely to revisit for the reduced complexity ability to understand at-a-glance.